src/AggregateFunctions/AggregateFunctionGroupArrayIntersect.cpp

```diff
@@ -0,0 +1,440 @@
+                        new_set.emplace(ArenaKeyHolder{data_column->getDataAt(offset + i), *arena}, it, inserted);
+                }
+                else
+                {
+                    const char * begin = nullptr;
+                    StringRef serialized = data_column->serializeValueIntoArena(offset + i, *arena, begin);
+                    assert(serialized.data != nullptr);
+                    it = set.find(serialized);
+
+                    if (it != nullptr)
+                        new_set.emplace(SerializedKeyHolder{serialized, *arena}, it, inserted);
+                }
+            }
+            set = std::move(new_set);
+        }
+    }
+
+    void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs, Arena * arena) const override
+    {
+        auto & set = this->data(place).value;
+        const auto & rhs_value = this->data(rhs).value;
+
+        if (this->data(rhs).version == 0)
+            return;
+
+        UInt64 version = this->data(place).version++;
+        if (version == 0)
+        {
+            bool inserted;
+            State::Set::LookupResult it;
+            for (auto & rhs_elem : rhs_value)
+            {
+                set.emplace(ArenaKeyHolder{rhs_elem.getValue(), *arena}, it, inserted);
+            }
+        }
+        else if (set.size() > 0)
+        {
+            auto create_new_map = [](auto & lhs_val, auto & rhs_val)
+            {
+                typename State::Set new_map;
+                for (auto & lhs_elem : lhs_val)
+                {
+                    auto val = rhs_val.find(lhs_elem.getValue());
+                    if (val != nullptr)
+                        new_map.insert(lhs_elem.getValue());
+                }
+                return new_map;
+            };
+            auto new_map = rhs_value.size() < set.size() ? create_new_map(rhs_value, set) : create_new_map(set, rhs_value);
+            set = std::move(new_map);
+        }
+    }
+
+    void serialize(ConstAggregateDataPtr __restrict place, WriteBuffer & buf, std::optional<size_t> /* version */) const override
+    {
+        auto & set = this->data(place).value;
+        auto & version = this->data(place).version;
+        writeVarUInt(version, buf);
+        writeVarUInt(set.size(), buf);
+
+        for (const auto & elem : set)
+            writeStringBinary(elem.getValue(), buf);
+    }
+
+    void deserialize(AggregateDataPtr __restrict place, ReadBuffer & buf, std::optional<size_t> /* version */, Arena * arena) const override
+    {
+        auto & set = this->data(place).value;
+        auto & version = this->data(place).version;
+        size_t size;
+        readVarUInt(version, buf);
+        readVarUInt(size, buf);
+        set.reserve(size);
+        UInt64 elem_version;
+        for (size_t i = 0; i < size; ++i)
+        {
+            auto key = readStringBinaryInto(*arena, buf);
+            readVarUInt(elem_version, buf);
+            if (elem_version == version)
```

Comment:
Can it be some other `version`? (I should probably look at the `Arena` implementation).